src/Sentry/Internal/BackpressureMonitor.cs

```diff
@@ -0,0 +1,166 @@
+    }
+
+    internal void IncrementDownsampleLevel()
+    {
+        var oldValue = _downsampleLevel;
+        if (oldValue < MaxDownsamples)
+        {
+            var newValue = oldValue + 1;
+            if (Interlocked.CompareExchange(ref _downsampleLevel, newValue, oldValue) == oldValue)
+            {
+                _logger?.LogDebug("System is under pressure, increasing downsample level to {0}.", newValue);
+            }
+        }
+    }
+
+    /// <summary>
+    /// A multiplier that can be applied to the SampleRate or TracesSampleRate to reduce the amount of data sent to
+    /// Sentry when the system is under pressure.
+    /// </summary>
+    public double DownsampleFactor
+    {
+        get
+        {
+            var level = _downsampleLevel;
+            return 1d / (1 << level); // 1 / (2^level) = 1, 1/2, 1/4, 1/8, ...
+        }
+    }
+
+    public void RecordRateLimitHit(DateTimeOffset when) => Interlocked.Exchange(ref _lastRateLimitEvent, when.Ticks);
+
+    public void RecordQueueOverflow() => Interlocked.Exchange(ref _lastQueueOverflow, _clock.GetUtcNow().Ticks);
+
+    private async Task DoWorkAsync(CancellationToken cancellationToken)
+    {
+        try
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                DoHealthCheck();
+
+                await Task.Delay(TimeSpan.FromSeconds(CheckIntervalInSeconds), cancellationToken).ConfigureAwait(false);
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Task was cancelled, exit gracefully
+        }
+    }
+
+    internal void DoHealthCheck()
+    {
+        if (IsHealthy)
+        {
+            var previous = Interlocked.Exchange(ref _downsampleLevel, 0);
+            if (previous > 0)
+            {
+                _logger?.LogDebug("System is healthy, resetting downsample level.");
+            }
+        }
+        else
+        {
+            IncrementDownsampleLevel();
+        }
+    }
+
+    /// <summary>
+    /// Checks for any recent queue overflows or any rate limit events since the last check.
+    /// </summary>
+    /// <returns></returns>
+    internal bool IsHealthy
+    {
+        get
+        {
+            var nowTicks = _clock.GetUtcNow().Ticks;
+            var recentOverflowCutoff = nowTicks - RecencyThresholdTicks;
+            var rateLimitCutoff = nowTicks - CheckIntervalTicks;
+            return LastQueueOverflowTicks < recentOverflowCutoff && LastRateLimitEventTicks < rateLimitCutoff;
+        }
+    }
+
+    public void Dispose()
+    {
+        try
+        {
+            _cts.Cancel();
+            _workerTask.Wait();
+        }
+        catch (AggregateException ex) when (ex.InnerException is OperationCanceledException)
+        {
+            // Ignore cancellation
+        }
+        catch (Exception ex)
+        {
+            // Log rather than throw
+            _logger?.LogWarning("Error in BackpressureMonitor.Dispose: {0}", ex);
```

Comment:
first arg should take Exception. The other overloads just call ToString on positional arguments
```suggestion
            _logger?.LogWarning(ex, "Error in BackpressureMonitor.Dispose");
```